lib/ansible/modules/storage/zfs/zpool.py

```diff
@@ -0,0 +1,454 @@
+    type: str
+  l2arc:
+    description:
+      - ZFS cache device or devices
+    type: str
+  state:
+    description:
+      - Create or delete the pool
+    choices: [ absent, present ]
+    required: true
+author:
+- Remy Mudingay
+- Stephane Armanet
+"""
+
+EXAMPLES = """
+- name: Create a new raidz zpool
+  zpool:
+    name: zfspool
+    devices:
+      - /dev/sdc
+      - /dev/sdd
+      - /dev/sde
+    raid_level: raidz
+    zil: mirror /dev/sdf /dev/sdg
+    vdev: 3
+    state: present
+
+- name: Create a new raid 0 stripe zpool
+  zpool:
+    name: zfspool
+    devices:
+      - /dev/sdb
+      - /dev/sdc
+      - /dev/sdd
+      - /dev/sde
+      - /dev/sdf
+    raid_level: raid0
+    l2arc: /dev/sdg
+    vdev: 5
+    ashift: 12
+    state: present
+
+- name: Create a new mirror zpool
+  zpool:
+    name: rpool
+    devices:
+      - /dev/sdb
+      - /dev/sdc
+      - /dev/sdd
+      - /dev/sde
+    raid_level: mirror
+    spare:
+      - /dev/sdf
+      - /dev/sdg
+    zil: mirror /dev/sdh /dev/sdi
+    l2arc: /dev/sdj
+    autoreplace: true
+    ashift: 12
+    vdev: 2
+    state: present
+
+- name: Create a new mirror zpool with a spare drive
+  zpool:
+    name: rpool
+    devices:
+      - /dev/sdc
+      - /dev/sdd
+    raid_level: mirror
+    vdev: 2
+    autoreplace: true
+    autoexpand: true
+    spare:
+      - /dev/sde
+    state: present
+
+- name: Add devices to an existing zpool
+  zpool:
+    name: rpool
+    add: true
+    devices:
+    - /dev/sdf
+    - /dev/sdg
+    raid_level: mirror
+    autoexpand: true
+    vdev: 2
+    state: present
+
+- name: Add spare dev to an existing zpool
+  zpool:
+    name: rpool
+    add: true
+    autoreplace: true
+    spare:
+    - /dev/sdf
+    state: present
+
+- name: Set options to an existing zpool
+  zpool:
+    name: zpool
+    sets: true
+    autoreplace: true
+    state: present
+
+- name: Set options to an existing zpool
+  zpool:
+    name: zpool
+    sets: true
+    autoexpand: off
+    state: present
+
+- name: Destroy an existing zpool
+  zpool:
+    name: rpool
+    state: absent
+"""
+
+RETURN = """ # """
+
+import os
+import re
+import ntpath
+import subprocess
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+class Zpool(object):
+
+    def __init__(self, module, name, state, raid_level, devices, spare, add, vdev, ashift, sets, autoreplace, autoexpand, zil, l2arc):
+        self.module = module
+        self.name = name
+        self.raid_level = raid_level
+        self.devices = devices
+        self.spare = spare
+        self.state = state
+        self.add = add
+        self.vdev = vdev
+        self.ashift = ashift
+        self.sets = sets
+        self.autoreplace = autoreplace
+        self.autoexpand = autoexpand
+        self.zil = zil
+        self.l2arc = l2arc
+        self.changed = False
+        self.zpool_cmd = module.get_bin_path('zpool', True)
+        self.zpool_grep = module.get_bin_path('grep', True)
```

Comment:
Using grep/awk here is not a very good idea. Everything can be done in Python and it makes the code portable across all ZFS platforms (Linux, FreeBSD, Solaris, illumos). Please consider dropping this in favour of Python code.

Reply:
grep and awk functions are now performed using python code.